File: haskell/hssearch/src/SearchOptions.hs
{-# LANGUAGE Arrows, NoMonomorphismRestriction #-}
module SearchOptions (
    SearchOption(..)
  , getSearchOptions
  , getUsage
  , settingsFromArgs) where

import Data.Char (toLower)
import Data.List (isPrefixOf, sortBy)
import Data.List.Split (splitOn)
import Data.Maybe (isJust)
import qualified Data.Text as T
import Text.XML.HXT.Core

import FileUtility (normalizeExtension)
import SearchSettings

data SearchOption = SearchOption { long, short, desc :: String }
  deriving (Show, Eq)

searchOptionsPath :: [Char]
searchOptionsPath = "/Users/cary/src/git/xsearch/shared/searchoptions.xml"

atTag tag = deep (isElem >>> hasName tag)

text = getChildren >>> getText

strip :: String -> String
strip = T.unpack . T.strip . T.pack

getSearchOption = atTag "searchoption" >>>
  proc f -> do
    long <- getAttrValue "long" -< f
    short <- getAttrValue "short" -< f
    desc <- text -< f
    returnA -< SearchOption { long = long, short = short, desc = strip desc }

getSearchOptions :: IO [SearchOption]
getSearchOptions = runX (readDocument [withValidate no] searchOptionsPath
  >>> getSearchOption)

getUsage :: [SearchOption] -> String
getUsage searchOptions =
  "Usage:\n  hssearch [options] <startpath>\n\nOptions:\n" ++
  searchOptionsToString searchOptions

getOptStrings :: [SearchOption] -> [String]
getOptStrings searchOptions = 
  map formatOpts searchOptions
  where formatOpts SearchOption {long=l, short=""} = long l
        formatOpts SearchOption {long=l, short=s}  = shortAndLong s l
        long l = "--" ++ l
        shortAndLong s l = "-" ++ s ++ "," ++ long l

getOptDesc :: SearchOption -> String
getOptDesc SearchOption {desc=""} = error "No description for SearchOption"
getOptDesc SearchOption {desc=d} = d

sortSearchOption :: SearchOption -> SearchOption -> Ordering
sortSearchOption (SearchOption {long=l1, short=s1}) (SearchOption {long=l2, short=s2}) =
  compare (shortOrLong s1 l1) (shortOrLong s2 l2)
  where
    shortOrLong "" l = l
    shortOrLong s l = (map toLower s) ++ "@" ++ l

sortSearchOptions :: [SearchOption] -> [SearchOption]
sortSearchOptions searchOptions = sortBy sortSearchOption searchOptions

padString :: String -> Int -> String
padString s len | length s < len = s ++ take (len - length s) (repeat ' ')
                | otherwise      = s

searchOptionsToString :: [SearchOption] -> String
searchOptionsToString searchOptions = 
  unlines $ zipWith formatOptLine optStrings optDescs
  where
    sorted = sortSearchOptions searchOptions
    optStrings = getOptStrings sorted
    optDescs = map getOptDesc sorted
    longest = maximum $ map length optStrings
    formatOptLine o d = "  " ++ (padString o longest) ++ "  " ++ d

data ActionType = ArgActionType
                | FlagActionType
                | UnknownActionType
  deriving (Show, Eq)

type ArgAction = SearchSettings -> String -> SearchSettings
type FlagAction = SearchSettings -> SearchSettings

argActions :: [(String, ArgAction)]
argActions = [ ("in-archivefilepattern", \ss s -> ss {inArchiveFilePatterns = inArchiveFilePatterns ss ++ [s]})
             , ("in-dirpattern", \ss s -> ss {inDirPatterns = inDirPatterns ss ++ [s]})
             , ("in-ext", \ss s -> ss {inExtensions = inExtensions ss ++ (newExtensions s)})
             , ("in-filepattern", \ss s -> ss {inFilePatterns = inFilePatterns ss ++ [s]})
             , ("in-linesafterpattern", \ss s -> ss {inLinesAfterPatterns = inLinesAfterPatterns ss ++ [s]})
             , ("in-linesbeforepattern", \ss s -> ss {inLinesBeforePatterns = inLinesBeforePatterns ss ++ [s]})
             , ("linesafter", \ss s -> ss {linesAfter = read s})
             , ("linesaftertopattern", \ss s -> ss {linesAfterToPatterns = linesAfterToPatterns ss ++ [s]})
             , ("linesafteruntilpattern", \ss s -> ss {linesAfterUntilPatterns = linesAfterUntilPatterns ss ++ [s]})
             , ("linesbefore", \ss s -> ss {linesBefore = read s})
             , ("maxlinelength", \ss s -> ss {maxLineLength = read s})
             , ("out-archivefilepattern", \ss s -> ss {outArchiveFilePatterns = outArchiveFilePatterns ss ++ [s]})
             , ("out-dirpattern", \ss s -> ss {outDirPatterns = outDirPatterns ss ++ [s]})
             , ("out-ext", \ss s -> ss {outExtensions = outExtensions ss ++ (newExtensions s)})
             , ("out-filepattern", \ss s -> ss {outFilePatterns = outFilePatterns ss ++ [s]})
             , ("out-linesafterpattern", \ss s -> ss {outLinesAfterPatterns = outLinesAfterPatterns ss ++ [s]})
             , ("out-linesbeforepattern", \ss s -> ss {outLinesBeforePatterns = outLinesBeforePatterns ss ++ [s]})
             , ("search", \ss s -> ss {searchPatterns = searchPatterns ss ++ [s]})
             ]

newExtensions :: String -> [String]
newExtensions x | ',' `elem` x = map normalizeExtension $ removeBlank (splitOn "," x)
                | otherwise    = [normalizeExtension x]
  where removeBlank :: [String] -> [String]
        removeBlank = filter (/="")

flagActions :: [(String, FlagAction)]
flagActions = [ ("allmatches", \ss -> ss {firstMatch=False})
              , ("archivesonly", \ss -> ss {archivesOnly=True})
              , ("debug", \ss -> ss {debug=True})
              , ("dotiming", \ss -> ss {doTiming=True})
              , ("excludehidden", \ss -> ss {excludeHidden=True})
              , ("firstmatch", \ss -> ss {firstMatch=True})
              , ("help", \ss -> ss {printUsage=True})
              , ("includehidden", \ss -> ss {excludeHidden=False})
              , ("listdirs", \ss -> ss {listDirs=True})
              , ("listfiles", \ss -> ss {listFiles=True})
              , ("listlines", \ss -> ss {listLines=True})
              , ("multilinesearch", \ss -> ss {multiLineSearch=True})
              , ("noprintmatches", \ss -> ss {printResults=False})
              , ("norecursive", \ss -> ss {recursive=False})
              , ("nosearcharchives", \ss -> ss {searchArchives=False})
              , ("printmatches", \ss -> ss {printResults=True})
              , ("recursive", \ss -> ss {recursive=True})
              , ("searcharchives", \ss -> ss {searchArchives=True})
              , ("uniquelines", \ss -> ss {uniqueLines=True})
              , ("verbose", \ss -> ss {verbose=True})
              , ("version", \ss -> ss {printVersion=True})
              ]

shortToLong :: [SearchOption] -> String -> String
shortToLong opts "" = ""
shortToLong opts s | length s == 2 && head s == '-' = "--" ++ getLongForShort s
                   | otherwise = s
  where getLongForShort s = (long . head . filter (\so -> short so == tail s)) opts

settingsFromArgs :: [SearchOption] -> [String] -> SearchSettings
settingsFromArgs opts args = recSettingsFromArgs defaultSearchSettings $ map (shortToLong opts) args
  where recSettingsFromArgs :: SearchSettings -> [String] -> SearchSettings
        recSettingsFromArgs settings args =
          case args of
          [] -> settings
          a:as | "-" `isPrefixOf` a ->
            case getActionType (argName a) of
            ArgActionType -> recSettingsFromArgs ((getArgAction (argName a)) settings (head as)) (tail as)
            FlagActionType -> recSettingsFromArgs ((getFlagAction (argName a)) settings) as
            UnknownActionType -> error ("Unknown option: " ++ (argName a))
          [s] -> settings {startPath=s}
        getActionType :: String -> ActionType
        getActionType a = if (isArgAction a)
                          then ArgActionType
                          else if (isFlagAction a)
                               then FlagActionType
                               else UnknownActionType
        argName :: String -> String
        argName a = (dropWhile (=='-') a)
        getArgAction ::String -> ArgAction
        getArgAction a = snd $ head $ filter (\(x,y) -> a==x) argActions
        getFlagAction ::String -> FlagAction
        getFlagAction a = snd $ head $ filter (\(x,y) -> a==x) flagActions
        isArgAction ::String -> Bool
        isArgAction a = isJust $ lookup a argActions
        isFlagAction ::String -> Bool
        isFlagAction a = isJust $ lookup a flagActions


main :: IO ()
main = do
  searchOptions <- getSearchOptions
  -- print searchOptions
  putStrLn $ getUsage searchOptions
  let args1 = [] :: [String]
  putStrLn $ "\nsettingsFromArgs " ++ show args1 ++ ": " ++ show (settingsFromArgs searchOptions args1)

  let args2 = ["--debug", "--archivesonly", "."]
  putStrLn $ "\nsettingsFromArgs " ++ show args2 ++ ": " ++ show (settingsFromArgs searchOptions args2)

  let args3 = ["--search", "SearchPattern", "--search", "SearchPattern2", "."]
  putStrLn $ "\nsettingsFromArgs " ++ show args3 ++ ": " ++ show (settingsFromArgs searchOptions args3)

  let args4 = ["-s", "SearchPattern", "-x", "hs", "-x", "hi", "."]
  putStrLn $ "\nsettingsFromArgs " ++ show args4 ++ ": " ++ show (settingsFromArgs searchOptions args4)

File: swift/swiftsearch/swiftsearch/SearchOptions.swift
//
//  SearchOptions.swift
//  swiftsearch
//
//  Created by Cary Clark on 5/17/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation

class SearchOption: CustomStringConvertible {
    let short: String
    let long: String
    let desc: String

    init(short: String, long: String, desc: String) {
        self.short = short
        self.long = long
        self.desc = desc
    }

    func sortArg() -> String {
        if !short.isEmpty {
            return short.lowercased() + "@" + long
        }
        return long
    }

    var description: String {
        var s = "SearchOption("
        s += "short: \"\(short)\""
        s += ", long: \"\(long)\""
        s += ", desc: \"\(desc)\""
        s += ")"
        return s
    }
}

class SearchOptionsXmlParser: NSObject, XMLParserDelegate {
    var searchOptions = [SearchOption]()
    let searchOptionNodeName = "searchoption"
    let longAttributeName = "long"
    let shortAttributeName = "short"
    var element = ""
    var longName = ""
    var shortName = ""
    var desc = NSMutableString()

    func parseFile(_ filepath: String) -> [SearchOption] {
        if (FileManager.default.fileExists(atPath: filepath)) {
            let data: Data? = try? Data(contentsOf: URL(fileURLWithPath: filepath))
            let inputStream: InputStream? = InputStream(data: data!)
            let parser: XMLParser? = XMLParser(stream: inputStream!)
            if parser != nil {
                parser!.delegate = self
                parser!.parse()
            }
        } else {
            print("ERROR: filepath not found: \(filepath)")
        }
        return searchOptions
    }

    func parser(_ parser: XMLParser, didStartElement elementName: String,
        namespaceURI: String?, qualifiedName qName: String?,
        attributes attributeDict: [String : String]) {
        element = elementName
        if (elementName as NSString).isEqual(to: searchOptionNodeName) {
            if attributeDict.index(forKey: longAttributeName) != nil {
                longName = (attributeDict[longAttributeName]!)
            }
            if attributeDict.index(forKey: shortAttributeName) != nil {
                shortName = (attributeDict[shortAttributeName]!)
            }
            desc = NSMutableString()
            desc = ""
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if element == searchOptionNodeName {
            desc.append(string)
        }
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String,
        namespaceURI: String?, qualifiedName qName: String?) {
        if (elementName as NSString).isEqual(to: searchOptionNodeName) {
            if !desc.isEqual(nil) {
                let trimmedDesc = desc.trimmingCharacters(in: whitespace as CharacterSet)
                searchOptions.append(SearchOption(short: shortName,
                    long: longName, desc: trimmedDesc))
            }
        }
    }
}

open class SearchOptions {
    fileprivate var searchOptions = [SearchOption]()
    fileprivate var longArgDict: [String:String] = [:]

    init() {
        setSearchOptions()
    }

    fileprivate func setSearchOptions() {
        let parser = SearchOptionsXmlParser()
        searchOptions = parser.parseFile(Config.searchOptionsPath)
        searchOptions.sort(by: { $0.sortArg() < $1.sortArg() })
        for so in searchOptions {
            longArgDict[so.long] = so.long
            if !so.short.isEmpty {
                longArgDict[so.short] = so.long
            }
        }
    }

    fileprivate let argActionDict: [String: (String, SearchSettings) -> ()] = [
        "in-archiveext": { (s: String, ss: SearchSettings) -> () in
            ss.addInArchiveExtension(s)
        },
        "in-archivefilepattern": { (s: String, ss: SearchSettings) -> () in
            ss.inArchiveFilePatterns.append(Regex(s))
        },
        "in-dirpattern": { (s: String, ss: SearchSettings) -> () in
            ss.inDirPatterns.append(Regex(s))
        },
        "in-ext": { (s: String, ss: SearchSettings) -> () in
            ss.addInExtension(s)
        },
        "in-filepattern": { (s: String, ss: SearchSettings) -> () in
            ss.inFilePatterns.append(Regex(s))
        },
        "in-filetype": { (s: String, ss: SearchSettings) -> () in
            ss.inFilePatterns.append(Regex(s))
        },
        "in-linesafterpattern": { (s: String, ss: SearchSettings) -> () in
            ss.inLinesAfterPatterns.append(Regex(s))
        },
        "in-linesbeforepattern": { (s: String, ss: SearchSettings) -> () in
            ss.inLinesBeforePatterns.append(Regex(s))
        },
        "linesafter": { (s: String, ss: SearchSettings) -> () in
            ss.linesAfter = Int(s)!
        },
        "linesaftertopattern": { (s: String, ss: SearchSettings) -> () in
            ss.linesAfterToPatterns.append(Regex(s))
        },
        "linesafteruntilpattern": { (s: String, ss: SearchSettings) -> () in
            ss.linesAfterUntilPatterns.append(Regex(s))
        },
        "linesbefore": { (s: String, ss: SearchSettings) -> () in
            ss.linesBefore = Int(s)!
        },
        "maxlinelength": { (s: String, ss: SearchSettings) -> () in
            ss.maxLineLength = Int(s)!
        },
        "out-archiveext": { (s: String, ss: SearchSettings) -> () in
            ss.addOutArchiveExtension(s)
        },
        "out-archivefilepattern": { (s: String, ss: SearchSettings) -> () in
            ss.outArchiveFilePatterns.append(Regex(s))
        },
        "out-dirpattern": { (s: String, ss: SearchSettings) -> () in
            ss.outDirPatterns.append(Regex(s))
        },
        "out-ext": { (s: String, ss: SearchSettings) -> () in
            ss.addOutExtension(s)
        },
        "out-filepattern": { (s: String, ss: SearchSettings) -> () in
            ss.outFilePatterns.append(Regex(s))
        },
        "out-filetype": { (s: String, ss: SearchSettings) -> () in
            ss.outFilePatterns.append(Regex(s))
        },
        "out-linesafterpattern": { (s: String, ss: SearchSettings) -> () in
            ss.outLinesAfterPatterns.append(Regex(s))
        },
        "out-linesbeforepattern": { (s: String, ss: SearchSettings) -> () in
            ss.outLinesBeforePatterns.append(Regex(s))
        },
        "search": { (s: String, ss: SearchSettings) -> () in
            ss.searchPatterns.append(Regex(s))
        }
    ]

    fileprivate let boolFlagActionDict: [String: (Bool, SearchSettings) -> ()] = [
        "allmatches": { (b: Bool, ss: SearchSettings) -> () in
            ss.firstMatch = !b
        },
        "archivesonly": { (b: Bool, ss: SearchSettings) -> () in
            ss.archivesOnly = b
            if b { ss.searchArchives = true }
        },
        "debug": { (b: Bool, ss: SearchSettings) -> () in
            ss.debug = b
            if b { ss.verbose = true }
        },
        "excludehidden": { (b: Bool, ss: SearchSettings) -> () in
            ss.excludeHidden = b
        },
        "firstmatch": { (b: Bool, ss: SearchSettings) -> () in
            ss.firstMatch = b
        },
        "help": { (b: Bool, ss: SearchSettings) -> () in
            ss.printUsage = b
        },
        "includehidden": { (b: Bool, ss: SearchSettings) -> () in
            ss.excludeHidden = !b
        },
        "listdirs": { (b: Bool, ss: SearchSettings) -> () in
            ss.listDirs = b
        },
        "listfiles": { (b: Bool, ss: SearchSettings) -> () in
            ss.listFiles = b
        },
        "listlines": { (b: Bool, ss: SearchSettings) -> () in
            ss.listLines = b
        },
        "multilinesearch": { (b: Bool, ss: SearchSettings) -> () in
            ss.multiLineSearch = b
        },
        "noprintmatches": { (b: Bool, ss: SearchSettings) -> () in
            ss.printResults = !b
        },
        "norecursive": { (b: Bool, ss: SearchSettings) -> () in
            ss.recursive = !b
        },
        "nosearcharchives": { (b: Bool, ss: SearchSettings) -> () in
            ss.searchArchives = !b
        },
        "printmatches": { (b: Bool, ss: SearchSettings) -> () in
            ss.printResults = b
        },
        "recursive": { (b: Bool, ss: SearchSettings) -> () in
            ss.recursive = b
        },
        "searcharchives": { (b: Bool, ss: SearchSettings) -> () in
            ss.searchArchives = b
        },
        "uniquelines": { (b: Bool, ss: SearchSettings) -> () in
            ss.uniqueLines = b
        },
        "verbose": { (b: Bool, ss: SearchSettings) -> () in
            ss.verbose = b
        },
        "version": { (b: Bool, ss: SearchSettings) -> () in
            ss.printVersion = b
        }
    ]

    func settingsFromArgs(_ args: [String], error: NSErrorPointer) -> SearchSettings {
        var i = 0
        let settings = SearchSettings()
        while i < args.count {
            var arg = args[i]
            if arg.hasPrefix("-") {
                while arg.hasPrefix("-") && arg.lengthOfBytes(using: String.Encoding.utf8) > 1 {
                    arg = String(arg[arg.index(arg.startIndex, offsetBy: 1)...])
                }
                if longArgDict.index(forKey: arg) != nil {
                    let longArg = longArgDict[arg]
                    if argActionDict.index(forKey: longArg!) != nil {
                        if args.count > i+1 {
                            argActionDict[longArg!]!(args[i+1], settings)
                            i += 1
                        } else {
                            setError(error, msg: "Missing argument for option \(arg)")
                        }
                    } else if boolFlagActionDict.index(forKey: longArg!) != nil {
                        boolFlagActionDict[longArg!]!(true, settings)
                    } else {
                        setError(error, msg: "Invalid option: \(arg)")
                    }
                } else {
                    setError(error, msg: "Invalid option: \(arg)")
                }
                
            } else {
                settings.startPath = args[i]
            }
            i += 1
        }
        return settings
    }

    func usage(_ code: Int32 = 0) {
        logMsg(getUsageString())
        exit(code)
    }

    func getUsageString() -> String {
        var s = "\nUsage:\n swiftsearch [options] -s <searchpattern> <startpath>\n\n"
        s += "Options:\n"
        let optStrings = searchOptions.map
            { $0.short.isEmpty ? "--\($0.long)" : "-\($0.short),--\($0.long)" }
        let optDescs = searchOptions.map { $0.desc }
        let longest = optStrings.map({ $0.lengthOfBytes(using: String.Encoding.utf8) }).max()!
        for i in 0 ..< optStrings.count {
            var optLine = " \(optStrings[i])"
            while optLine.lengthOfBytes(using: String.Encoding.utf8) <= longest {
                optLine += " "
            }
            optLine += "  \(optDescs[i])\n"
            s += optLine
        }
        return s
    }
}
